play-services-base/src/main/java/org/microg/gms/common/MultiConnectionKeeper.java

```diff
@@ -21,32 +21,142 @@
 import android.content.Context;
 import android.content.Intent;
 import android.content.ServiceConnection;
+import android.content.SharedPreferences;
 import android.content.pm.ApplicationInfo;
 import android.content.pm.PackageManager;
 import android.content.pm.PermissionInfo;
 import android.content.pm.ResolveInfo;
+import android.content.pm.Signature;
 import android.os.IBinder;
 import android.util.Log;
 
+import java.security.MessageDigest;
+import java.security.NoSuchAlgorithmException;
+import java.util.Arrays;
 import java.util.HashMap;
 import java.util.HashSet;
+import java.util.LinkedList;
+import java.util.List;
 import java.util.Map;
 import java.util.Set;
 
 import static android.os.Build.VERSION.SDK_INT;
 import static android.os.Build.VERSION_CODES.ICE_CREAM_SANDWICH;
 import static org.microg.gms.common.Constants.GMS_PACKAGE_NAME;
+import static org.microg.gms.common.Constants.MICROG_PACKAGE_NAME;
+import static org.microg.gms.common.Constants.GMS_PACKAGE_SIGNATURE_SHA1;
+import static org.microg.gms.common.Constants.GMS_SECONDARY_PACKAGE_SIGNATURE_SHA1;
+import static org.microg.gms.common.Constants.MICROG_PACKAGE_SIGNATURE_SHA1;
 
 public class MultiConnectionKeeper {
     private static final String TAG = "GmsMultiConKeeper";
+    private static final String PREF_MASTER = "GmsMultiConKeeper";
+    private static final String PREF_TARGET = "GmsMultiConKeeper_target";
```

Comment:
The preference name is already scoped through the file name (`PREF_MASTER`), so for readability I think it would be preferably to not prefix the preference name with the file name.